### ISSUE TITLE
Fix label filtering bug

### ISSUE THREAD

**Issue by nathanielrindlaub** [tnc-ca-geo/animl-api#43]
images with non-validated labels within locked objects are still passing the labels filter. For example, if you only have the 'rodent' label selected, it's returning locked objects that have a different label validated (e.g. skunk, bird, etc.) and the rodent label validation = null. There must be something wrong with the query [here](https://github.com/tnc-ca-geo/animl-api/blob/1ede0b94d9a7c5e55c1de50ce70d9f04e5af1203/src/api/db/models/utils.js#L84).

**Comment by nathanielrindlaub** [comment]
Appears to still have issues. Investigate. Fox label is especially problematic.

**Comment by nathanielrindlaub** [comment]
Also "none" filter seems buggy?

**Comment by nathanielrindlaub** [comment]
I did a spot check on the images that are incorrectly passing through the label filter, and it appears it's due to the problem I  described in the comments above (if an object is locked and has _any_ validated labels that match the label filters, it will pass, and an object could have multiple validated labels. We only want to assess the _first_ validated label in the labels array). 

In order to fix this, I think we're going to have to move to using MongoDB aggregation pipelines rather than find queries. mongo-cursor-pagination [should](https://github.com/mixmaxhq/mongo-cursor-pagination/blob/master/src/aggregate.js#L5) support this, but requires testing. 

The following changes would need to be made: 

1. Refactor existing queries to use aggregation pipeline:
- [ ] instead of conditionality composing a bunch of MongoDB query objects into one large query in `buildFilter()` (`src/api/db/models/utils.js`), test building composable `$match` stages that get assembled into an aggregation pipeline.
- [ ] update `countImages` in the Image model to use Mongoose's `Model.aggregate()` function instead of `Model.where()`. Might also have to use `[Model.aggregate().count()](https://mongoosejs.com/docs/api/aggregate.html#aggregate_Aggregate-count)` instead of `Model.query.countDocuments()` but worth testing first.
- [ ] update `queryByFilter` in the Image model to use mongo-cursor-pagination's [aggregation wrapper](https://github.com/mixmaxhq/mongo-cursor-pagination/blob/master/src/aggregate.js) instead of it's Mongoose plugin (it looks like the plugin only supports `search()` or `find()`, not `aggregate()`.

2. Update labels filter aggregation stage 
- [ ] if user passes in a labels filter array, it should append two stages to the aggregation pipeline: the first being a `$set` stage, in which we use `$filter` with a`limit` option of 1 to add a field to all objects that's just the first validated label of the labels array, and then a second `$match` stage to filter out the correct image documents. I think something like the following should work:

```
[{
    $set: {
        "objects.firstValidLabel": {
            $filter: {
                input: "$objects.labels",
                cond: { "$label.validation.validated": true },
                as: "label",
                limit: 1
            }
        }
    }
}, {
    $match: {$or: [

        // has an object that is locked,
        // and it's first validated label is included in filters
        {objects: {$elemMatch: {
            locked: true,
            "firstValidLabel.category": {$in: labels},
         }}},

        // has an object is not locked, but it has label that is 
        // not-invalidated and included in filters
        {objects: {$elemMatch: {
            locked: false,
            labels: {$elemMatch: {
                'validation.validated': {$not: {$eq: false}},
                category: {$in: labels},
            }}
        }}},
    ]};
}}]
```

**Comment by nathanielrindlaub** [comment]
Unfortunately, the key feature we were leaning on for a solution here (`$filter`'s `limit` option), is a new feature only available in MongoDB 5.3. I will look into upgrading...
